### ISSUE TITLE
Can't merge hierarchical options into submodule (or freeform option)

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#146882]
### Describe the bug

Part of an option tree may be free-form.

Extending the free-form part in a separate module is hard and prevents certain use cases and migrations.

### Steps To Reproduce

```nix
let
  lib = import ./lib;
  inherit (lib) types mkOption;
in
(lib.evalModules {
  modules = [
    {
      options.free = mkOption {
        type = lib.types.submoduleWith {
          modules = [
            # UPDATE: A freeformType is not necessarily part of the problem.
            # { freeformType = lib.types.anything; }
          ];
        };
        default = {};
      };
    }
    {
      options.free = mkOption {
        type = lib.types.submoduleWith {
          modules = [
            {
              options.complicated = mkOption {
                type = types.int;
                default = 1;
              };
            }
          ];
        };
      };
    }
    {
      options.free.specific = mkOption {
        type = lib.types.int;
        default = 2;
      };
    }
  ];
}).config
```

```console
$ nix-instantiate test.nix  --eval --strict
error: The option `free' in `<unknown-file>' is a prefix of options in `<unknown-file>'.
(use '--show-trace' to show detailed location information)

```

The submodule only exists in order to provide a freeform type for that subtree, but it is conceptually still the same tree.

### Expected behavior

```console
$ nix-instantiate test.nix  --eval --strict
{ free = { complicated = 1; specific = 2; }; }
```

### Additional context

A semi-workaround is to make writing the unnecessary module boundary easier.

```
    mkSubmoduleOptions =
      options:
      mkOption {
        type = types.submoduleWith {
          modules = [ { inherit options; } ];
        };
      };
```

However, this is needlessly complicated and it does not help in situations where a freeform type is introduced in a decentralized scenario.

### Notify maintainers

@Infinisil 

### Metadata

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - lib.evalModules
  - lib.mkOption
# a list of nixos modules affected by the problem
module:
```

**Comment by Infinisil** [comment]
> I would prefer for `options.foo."*".x` to just create `x` in the submodule instance where `name` is `"*"`

Isn't that the same behavior as assigning `config.foo."*".x`? Or do you mean something different

**Comment by roberth** [comment]
Here `x` is an option declaration, not just a config definition.

I actually know a use case for this, which I've encountered in the pre-commit-hooks.nix project, where you can generally do everything in the part of the tree that builds the config file, but some hooks have interesting options that you want to model. These are currently unintuitive to write and I don't know if they'd even end up in the docs. It's like `config.systemd.services` vs `config.services`, except the balance is such that you rarely need `services`. Here's an example in terms of `systemd`, so you don't need to dive into pre-commit-hooks.nix. _I don't think we want this for systemd, because we _do_ want to abstract over its details._ Anyway, it goes like this:

```nix
{
  config.systemd.services.foo = { ... }: {
    options.fooSpecific = lib.mkOption {};
  };
}
```

whereas more natural would be

```nix
{
  options.systemd.services.foo.fooSpecific = lib.mkOption {};
}
```

**Comment by roberth** [comment]
@Infinisil I forgot about this limitation today. We should be able to do this: https://github.com/NixOS/nixpkgs/pull/155883

**Comment by roberth** [comment]
Possible fix in #156533
